Add per-network config file network.conf

ajtowns opened the pull request:
Adds a per-net network.conf as per #9374 ; this is pretty much the same as #9402 which was closed for lack of interest.

It's somewhat interesting in the context of #10994 because it provides somewhere to stick the network-specific vbignore settings. Unless I'm missing something, I don't think #10267 provides an easy way of doing per-network config files, so wouldn't work well for that particular use case.

This patch doesn't include any tests, so is marked as WIP. There's some test code in #10267 that would presumably be able to be adapted pretty easily.

laanwj commented:
Concept ACK

laanwj commented on doc/man/bitcoin-cli.1:
```diff
@@ -22,6 +22,10 @@ This help message
 .IP
 Specify configuration file (default: bitcoin.conf)
 .HP
+\fB\-netconf=\fR<file>
```
I wonder - is there any practical point in being able to override this?

MeshCollider replied:
Aren't the doc/man/ files automatically generated anyway so shouldn't be modified directly?

ajtowns replied:
@MeshCollider Yep, that's correct; worse the help text for bitcoin-cli is missing. Will fix.

ajtowns commented on doc/man/bitcoin-cli.1:
```diff
@@ -22,6 +22,10 @@ This help message
 .IP
 Specify configuration file (default: bitcoin.conf)
 .HP
+\fB\-netconf=\fR<file>
```
It could be useful if you wanted to temporarily restart bitcoind without using the network.conf file. You could have a bitcoin-alt.conf that specified netconf=network-alt.conf so that you could load bitcoind with -conf=bitcoin-alt.conf and have that load network specific alternative configurations too. Not sure how realistic/valuable those are, though. Not having the option would be inconsistent; and having the option makes the name of the config file at least a little discoverable...

ajtowns commented:
Commit 7a71b09 adds basic functional tests; it makes use of -netconf being a modifiable parameter to point bitcoind at different files in different ways and checks that has different results.

Commit 8b48aa5 cleans up the code a little, but also allows you to specify an absolute path for network.conf if you want to for some reason (not that doing so actually makes a lot of sense).

Commit 70f5388 removes the default for fNetSpecific for calls to Get/ReadConfigFile, which seems sensible since the default would need to be false to be backwards compatible with the old calls for bitcoin.conf, but is the opposite to the default for the same parameter for GetDataDir().

ajtowns commented:
One possibly confusing behaviour: possible to set "-testnet" or "-regtest" in a network.conf where those parameters would be wrong; if you do so, they won't take effect or cause problems because ChainNameFromCommandLine() has already been called and isn't called again; but if some future change to the code does try to call ChainNameFromCommandLine() sometime after startup, and a network.conf sets a wrong parameter, it could blow up in your face.

I'm not sure if this is worth fixing...

(Adding a "static const string *result;" to ChainNameFromCommandLine() to memoise it at first call, and just reuse the result without reexamining the arguments later seems to be sufficient to fix the problem if it is worth fixing. Giving an error if regtest/testnet is specified in network.conf would probably be better, but seems like it violates the abstraction layer a fair bit...)

jnewbery commented:
I've wanted to rebase and revive #9402 for some time, but I was waiting for #10267, both because this conflicts heavily with that PR, and also because I think once #10267 is merged, then this PR becomes more straightforward.

ajtowns commented:
@jnewbery Hmm, looking at #10267 in a bit more detail, the current patch seems to only allow a single includeconf directive to have an effect anyway, rather than multiple includes or recursive includes; so you could actually get the same effect as that patch with this one by just giving an absolute path to the netconf directory here. It doesn't look to me like there's really much difficulty in combining the two patches either once one or the other gets merged (I'm not seeing much room for either to get much simpler, though -- you still have to finish loading bitcoin.conf then load the chain params to work out which network you're on before you can load the right network.conf).

ajtowns commented:
Squashed most of the commits, and rebased to current master.

ajtowns committed:
Add per-network config file network.conf

ajtowns committed:
Be explicit whether config file is net specific

Remove the default for whether GetConfigFile and ReadConfigFile
expects the file to be network specific or not, requiring true
or false to be passed in for fNetworkSpecific on any call.

ajtowns commented:
Rebased to current master. Also: no longer committing changes to manpages, functional test updated due to API changes on master, help text for bitcoin-cli --netconf added, and minor code style fixes.

ajtowns commented:
Rebased to fix test_runner conflicts

ajtowns commented:
@MeshCollider In normal usage, -netconf isn't needed -- you just add the network.conf file to .bitcoin/testnet3/ (or wherever) if you want to use it, and you don't if you don't want to use it. The main benefits to having the option are (a) it makes it easier to test (can just specify different config files to check different behaviour, rather than having to move files around), and (b) it means it shows up when you say --help so people have some chance of knowing it exists. I think that's enough to justify having the option, but happy to drop it if people think that makes it simpler.

MeshCollider commented:
Upon discussion on IRC: I suggested using network.conf if `-conf` is not specified but using the `-conf` if it is and not using network.conf at all. This removes the need for the netconf argument. Reasoning is that is a user specified a config file they probably want to use it on the network they specify as well, so we really only need to change the default behaviour. But unsure if that's too limiting

sipa commented:
Concept ACK

ajtowns commented:
@MeshCollider see https://github.com/ajtowns/bitcoin/pull/2 . not a big fan of repeating the conditionals still, but it's not too bad.

promag commented on src/util.h:
```diff
@@ -200,7 +201,7 @@ class ArgsManager
 public:
     void ParseParameters(int argc, const char*const argv[]);
-    void ReadConfigFile(const std::string& confPath);
+    void ReadConfigFile(const std::string& confPath, bool fNetSpecific);
```
Same as above.

promag commented on src/util.h:
```diff
@@ -171,7 +172,7 @@ bool TryCreateDirectories(const fs::path& p);
 const fs::path &GetDataDir(bool fNetSpecific = true);
 void ClearDatadirCache();
-fs::path GetConfigFile(const std::string& confPath);
+fs::path GetConfigFile(const std::string& confPath, bool fNetSpecific);
```
Set default value `bool fNetSpecific = false`? In that case remove changes where it's false.

ajtowns replied:
The "Be explicit whether config file is net specific" commit removes the fNetSpecific=false default argument, because otherwise it would have the opposite default to GetDataDir() which seems confusing to me.

ajtowns replied:
Err, the implication being that just dropping/not merging that commit is how you'd get the fNetSpecific=false default.

promag reviewed (commented):
Meanwhile some nits.

ajtowns commented on test/functional/netconf.py:
```diff
@@ -0,0 +1,82 @@
+import os
+from test_framework.test_framework import BitcoinTestFramework
+
+class IncludeConfTest(BitcoinTestFramework):
```
Changed.

ajtowns commented on test/functional/netconf.py:
```diff
@@ -0,0 +1,82 @@
+
+        # otherwise loads specified network conf
+        self.stop_node(0)
+        self.start_node(0, ["-conf=bitcoin-altnet.conf"])
```
Changed.

ajtowns commented on test/functional/netconf.py:
```diff
@@ -0,0 +1,82 @@
+
+        # netconf on command line overrides specification in bitcoin.conf
+        self.stop_node(0)
+        self.start_node(0, ["-conf=bitcoin-altnet.conf", "-netconf=network.conf"])
```
Changed.

ajtowns commented on test/functional/netconf.py:
```diff
@@ -0,0 +1,82 @@
+
+        # if file doesn't exist, everything is fine
+        self.stop_node(0)
+        self.start_node(0, ["-netconf=no-network.conf"])
```
Changed.

ajtowns commented on test/functional/netconf.py:
```diff
@@ -0,0 +1,82 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Tests the network-specific config file
```
Changed.

ajtowns commented:
Rebased, small improvements to test case, nit fixes.

ajtowns commented:
Rebased to fix test_runner.py conflict, renamed test case to conform to naming convention of #11796

MeshCollider commented:
This is now superseded by https://github.com/bitcoin/bitcoin/pull/11862

ajtowns commented:
Abandoning in favour of #11862
